### PR TITLE
Fixes false positives with public methods which are not allowed

### PR DESCRIPTION
**Pull request by 4creators**
as test methods in xunit: "public static" and "public abstract"

PublicMethodShouldBeMarkedAsTest greedily issues warning for 'public static' and 'public abstract'
methods although they can not be used as a test method in xunit framework. Introduced code changes
verify if public method is abstract or is static and if detected skip further checks.

Fixes https://github.com/xunit/xunit/issues/1466

### CODE REVIEW

**Comment by 4creators** [comment]
PTAL @marcind

File: src/xunit.analyzers.vsix/source.extension.vsixmanifest [hunk 1]
```diff
@@ -1,21 +1,21 @@
...
+        <Dependency Id="Microsoft.Framework.NDP" DisplayName="Microsoft .NET Framework" d:Source="Manual" Version="[4.5,)" />
+    </Dependencies>
+    <Assets>
+        <Asset Type="Microsoft.VisualStudio.MefComponent" d:Source="Project" d:ProjectName="xunit.analyzers" Path="|xunit.analyzers|"/>
+        <Asset Type="Microsoft.VisualStudio.Analyzer" d:Source="Project" d:ProjectName="xunit.analyzers" Path="|xunit.analyzers|"/>
+    </Assets>
+    <Prerequisites>
+        <Prerequisite Id="Microsoft.VisualStudio.Component.CoreEditor" Version="[15.0.26208.0,16.0)" DisplayName="Visual Studio core editor" />
+    </Prerequisites>
 </PackageManifest>
```
Comment (bradwilson) [inline src/xunit.analyzers.vsix/source.extension.vsixmanifest]: Please do not reformat files. The entire file looks changed because of spacing differences.

File: src/xunit.analyzers/PublicMethodShouldBeMarkedAsTest.cs [hunk 2]
```diff
@@ -42,7 +43,7 @@ internal override void AnalyzeCompilation(CompilationStartAnalysisContext compil
                     symbolContext.CancellationToken.ThrowIfCancellationRequested();
 
                     var method = (IMethodSymbol)member;
-                    if (method.MethodKind != MethodKind.Ordinary)
+                    if (method.MethodKind != MethodKind.Ordinary || method.IsStatic || method.IsAbstract)
```
Comment (bradwilson) [inline src/xunit.analyzers/PublicMethodShouldBeMarkedAsTest.cs]: xUnit.net does support public static test methods.

**Review by bradwilson — CHANGES_REQUESTED** [review]

File: src/xunit.analyzers/PublicMethodShouldBeMarkedAsTest.cs (hunk 2 above)
Comment (4creators) [reply src/xunit.analyzers/PublicMethodShouldBeMarkedAsTest.cs]: Yeah, I have tested initially and xunit failed to detect that but using xunit.analyzers.tests I got it as You said. My wrong

File: src/xunit.analyzers.vsix/source.extension.vsixmanifest [hunk 3]
```diff
@@ -1,7 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <PackageManifest Version="2.0.0" xmlns="http://schemas.microsoft.com/developer/vsx-schema/2011" xmlns:d="http://schemas.microsoft.com/developer/vsx-schema-design/2011">
   <Metadata>
-    <Identity Id="xunit.analyzers..d09124d1-9c5e-40de-b1de-35120196c3e0" Version="1.0" Language="en-US" Publisher="Marcin Dobosz"/>
+    <Identity Id="xunit.analyzers..d09124d1-9c5e-40de-b1de-35120196c3e0" Version="1.0.1" Language="en-US" Publisher="Marcin Dobosz"/>
```
Comment (marcind) [inline src/xunit.analyzers.vsix/source.extension.vsixmanifest]: Why is this file changed at all? Please undo.

Comment (4creators) [reply src/xunit.analyzers.vsix/source.extension.vsixmanifest]: Thought that xunit has semantic versioning. Change in error/warning detection syntax changes behaviour so I have incremented build version change.

Obviously will change it as requested.

**Comment by 4creators** [comment]
Addressed all review remarks 
@bradwilson @marcind

File: src/xunit.analyzers/PublicMethodShouldBeMarkedAsTest.cs [hunk 4]
```diff
@@ -42,7 +43,7 @@ internal override void AnalyzeCompilation(CompilationStartAnalysisContext compil
                     symbolContext.CancellationToken.ThrowIfCancellationRequested();
 
                     var method = (IMethodSymbol)member;
-                    if (method.MethodKind != MethodKind.Ordinary)
+                    if (method.MethodKind != MethodKind.Ordinary || method.IsAbstract)
```
Comment (jamesqo) [inline src/xunit.analyzers/PublicMethodShouldBeMarkedAsTest.cs]: Is this ever true? We returned earlier if `type.IsAbstract` was true, so logically this must be a concrete type and none of its members can be abstract.

Comment (jamesqo) [reply src/xunit.analyzers/PublicMethodShouldBeMarkedAsTest.cs]: Maybe the best fix would be to remove the `type.IsAbstract` check above and keep this as-is.

File: test/xunit.analyzers.tests/PublicMethodShouldBeMarkedAsTestTests.cs [hunk 5]
```diff
@@ -38,6 +39,18 @@ public class PublicMethodShouldBeMarkedAsTestTests
         }
 
         [Fact]
+        public async void DoesNotFindErrorForPublicAbstractMethod()
+        {
+            var diagnostics = await CodeAnalyzerHelper.GetDiagnosticsAsync(analyzer,
+@"public abstract class TestClass {
+    [Xunit.Fact] public void TestMethod() { }
+    public abstract void AbstractTestError();
```
Comment (jamesqo) [inline test/xunit.analyzers.tests/PublicMethodShouldBeMarkedAsTestTests.cs]: Nit: Rename to `AbstractMethod`, since it isn't a test method and it doesn't raise an error.

File: test/xunit.analyzers.tests/PublicMethodShouldBeMarkedAsTestTests.cs [hunk 6]
```diff
@@ -38,6 +39,18 @@ public class PublicMethodShouldBeMarkedAsTestTests
...
+}");
+
+            Assert.Empty(diagnostics);
+        }
+
```
Comment (jamesqo) [inline test/xunit.analyzers.tests/PublicMethodShouldBeMarkedAsTestTests.cs:52]: Would be good to also add this test:

```cs
        [Fact]
        public async void DoesNotFindErrorForPublicAbstractMethodMarkedWithFact()
        {
            var diagnostics = await CodeAnalyzerHelper.GetDiagnosticsAsync(analyzer,
@"public abstract class TestClass {
    [Xunit.Fact] public void TestMethod() { }
    [Xunit.Fact] public abstract void AbstractTestMethod();
}");
            Assert.Empty(diagnostics);
        }
```

I can't test right now but that should work fine at runtime, since the `[Fact]` attribute is inherited on overriding members.

File: src/xunit.analyzers/PublicMethodShouldBeMarkedAsTest.cs (hunk 4 above)
Comment (4creators) [reply src/xunit.analyzers/PublicMethodShouldBeMarkedAsTest.cs]: Sorry for late replay, I was stuck with other work for a while.
This way it is set it can skip diagnostics in partially invalid code during code editing i.e.  concrete class with abstract method defined or skip analysis in abstract class for public concrete methods.

File: test/xunit.analyzers.tests/PublicMethodShouldBeMarkedAsTestTests.cs (hunk 5 above)
Comment (4creators) [reply test/xunit.analyzers.tests/PublicMethodShouldBeMarkedAsTestTests.cs]: Will do

File: test/xunit.analyzers.tests/PublicMethodShouldBeMarkedAsTestTests.cs (hunk 6 above)
Comment (4creators) [reply test/xunit.analyzers.tests/PublicMethodShouldBeMarkedAsTestTests.cs:52]: OK it makes sense

**Comment by 4creators** [comment]
PTAL
@bradwilson @marcind @jamesqo 
I left one unresolved problem to decide see: https://github.com/xunit/xunit.analyzers/pull/82#discussion_r144719822

File: src/xunit.analyzers/PublicMethodShouldBeMarkedAsTest.cs (hunk 4 above)
Comment (jamesqo) [reply src/xunit.analyzers/PublicMethodShouldBeMarkedAsTest.cs]: OK, it's fine to leave this as-is then, but can you add a comment about it?

**Comment by jamesqo** [comment]
After you leave a comment LGTM

**Review by bradwilson — APPROVED** [review]
👍 once the requested comment has been made

**Comment by 4creators** [comment]
OK All feedback addressed 😄 
PTAL @bradwilson

**Comment by marcind** [comment]
Thanks for the submission
